File: ZezeJava/ZezeJava/src/main/java/Zeze/Transaction/Collections/LogBean.java
package Zeze.Transaction.Collections;

import Zeze.Serialize.ByteBuffer;
import Zeze.Serialize.IByteBuffer;
import Zeze.Transaction.Bean;
import Zeze.Transaction.Changes;
import Zeze.Transaction.Log;
import Zeze.Transaction.LogDynamic;
import Zeze.Transaction.Savepoint;
import Zeze.Util.IntHashMap;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class LogBean extends Log {
	private static final int TYPE_ID = Bean.hash32("Zeze.Transaction.Collections.LogBean");

	private final Bean self;
	private @Nullable IntHashMap<Log> variables;

	public LogBean(Bean belong, int varId, Bean self) {
		super(belong, varId);
		this.self = self;
	}

	@Override
	public Category category() {
		return Category.eHistory;
	}

	@Override
	public int getTypeId() {
		return TYPE_ID;
	}

	public final @Nullable IntHashMap<Log> getVariables() {
		return variables;
	}

	public final @NotNull IntHashMap<Log> getVariablesOrNew() {
		var variables = this.variables;
		if (variables == null)
			this.variables = variables = new IntHashMap<>();
		return variables;
	}

	public final Bean getThis() {
		return self;
	}

	@Override
	public void commit() {
		throw new UnsupportedOperationException();
	}

	// LogBean仅在_final_commit的Collect过程中创建，不会参与Savepoint。
	@Override
	public @NotNull Log beginSavepoint() {
		throw new UnsupportedOperationException();
	}

	@Override
	public void endSavepoint(@NotNull Savepoint currentSp) {
		throw new UnsupportedOperationException();
	}

	@Override
	public void encode(@NotNull ByteBuffer bb) {
		var vars = variables;
		//logger.info("LogBean.this=" + getThis().getClass().getName());
		if (vars != null) {
			bb.WriteUInt(vars.size());
			for (var it = vars.iterator(); it.moveToNext(); ) {
				Log log = it.value();
				bb.WriteInt4(log.getTypeId());
				bb.WriteUInt(log.getVariableId());
				log.encode(bb);
				//logger.info("key=" + it.key() + " typeId=" + log.getTypeId() + " varId=" + log.getVariableId() + " name=" + log.getClass().getName());
			}
		} else
			bb.WriteUInt(0);
	}

	@Override
	public void decode(@NotNull IByteBuffer bb) {
		int n = bb.ReadUInt();
		if (n > 0) {
			var variables = getVariablesOrNew();
			variables.clear();
			for (; n > 0; --n) {
				int typeId = bb.ReadInt4();
				int varId = bb.ReadUInt();
				Log log = create(typeId, varId);
				log.decode(bb);
				variables.put(varId, log);
			}
		} else if (variables != null)
			variables.clear();
	}

	// 仅发生在事务执行期间。decode-Apply不会执行到这里。
	@Override
	public void collect(@NotNull Changes changes, @NotNull Bean recent, @NotNull Log vlog) {
		if (getVariablesOrNew().putIfAbsent(vlog.getVariableId(), vlog) == null)
			changes.collect(recent, this); // 向上传递
	}

	@Override
	public @NotNull String toString() {
		var sb = new StringBuilder();
		ByteBuffer.BuildSortedString(sb, variables);
		return sb.toString();
	}

	public static void encodeLogBean(@NotNull ByteBuffer bb, @NotNull LogBean logBean) {
		// 使用byte，未来可能扩展其他LogBean子类。
		if (logBean instanceof LogDynamic)
			bb.WriteByte(1);
		else // 全部都是LogBean子类，只能else。
			bb.WriteByte(0);
		logBean.encode(bb);
	}

	public static @NotNull LogBean decodeLogBean(@NotNull IByteBuffer bb) {
		int type = bb.ReadByte();
		LogBean logBean;
		switch (type) {
		case 0:
			logBean = new LogBean(null, 0, null);
			break;
		case 1:
			logBean = new LogDynamic(null, 0, null);
			break;
		default:
			throw new RuntimeException("unknown logBean subclass type=" + type);
		}
		logBean.decode(bb);
		return logBean;
	}
}

File: Zege/Zeze/Transaction/Collections/LogBean.cs
using System.Collections.Generic;
using System.Text;
using Zeze.Serialize;
using Zeze.Util;

namespace Zeze.Transaction.Collections
{
    public class LogBean : Log
    {
        public static readonly string StableName = Reflect.GetStableName(typeof(LogBean));
        public static readonly int TypeId_ = FixedHash.Hash32(StableName);

        public override int TypeId => TypeId_;

        public Dictionary<int, Log> Variables { get; } = new Dictionary<int, Log>();

#if !USE_CONFCS
		public Bean This { get; set; }

		// LogBean仅在_final_commit的Collect过程中创建，不会参与Savepoint。
        internal override Log BeginSavepoint()
        {
            throw new System.NotImplementedException();
        }

        internal override void EndSavepoint(Savepoint currentsp)
        {
            throw new System.NotImplementedException();
        }

        // 仅发生在事务执行期间。Decode-Apply不会执行到这里。
        public override void Collect(Changes changes, Bean recent, Log vlog)
        {
            if (Variables.TryAdd(vlog.VariableId, vlog))
            {
                // 向上传递
                changes.Collect(recent, this);
            }
        }

        public override void Commit()
        {
            throw new System.NotImplementedException();
        }

        public static void EncodeLogBean(ByteBuffer bb, LogBean logBean)
        {
            // 使用byte，未来可能扩展其他LogBean子类。
            if (logBean is LogDynamic)
                bb.WriteByte(1);

            else // 全部都是LogBean子类，只能else。
                bb.WriteByte(0);
            logBean.Encode(bb);
        }
#endif

        public static LogBean DecodeLogBean(ByteBuffer bb)
        {
            var type = bb.ReadByte();
            LogBean logBean;
            switch (type)
            {
                case 0:
                    logBean = new LogBean();
                    break;
                case 1:
#if USE_CONFCS
                    logBean = new LogConfDynamic();
#else
                    logBean = new LogDynamic();
#endif
                    break;
                default:
                    throw new System.Exception("unknown logBean subclass type=" + type);
            }
            logBean.Decode(bb);
            return logBean;
        }

        public override void Decode(ByteBuffer bb)
        {
            for (int i = bb.ReadUInt(); i > 0; --i)
            {
                var typeId = bb.ReadInt4();
                var log = Create(typeId);

                var varId = bb.ReadUInt();
                log.VariableId = varId;
                log.Decode(bb);

                Variables.Add(varId, log);
            }
        }

        public override void Encode(ByteBuffer bb)
        {
            bb.WriteUInt(Variables.Count);
            foreach (var log in Variables.Values)
            {
                bb.WriteInt4(log.TypeId);
                bb.WriteUInt(log.VariableId);
                log.Encode(bb);
            }
        }

        public override string ToString()
        {
            var sb = new StringBuilder();
            ByteBuffer.BuildString(sb, Variables, ComparerInt.Instance);
            return sb.ToString();
        }
    }
}
